File: sdk/c#/PicarioXPO.RenderAPI.V2/OverlayKeyExtractors/NameOverlayKey.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web;

namespace PicarioXPO.RenderAPI.V2.OverlayKeyExtractors
{
    internal class NameOverlayKey : OverlayKey
    {
        public override string GetValues(IEnumerable<XpoUrlOverlay> overlays)
        {
            var urlOverlays = overlays as XpoUrlOverlay[] ?? overlays.ToArray();
            var max = urlOverlays.GetMaxOverlayNumber();

            for (var i = 0; i <= max; i++)
            {
                var urlOverlay = urlOverlays.FirstOrDefault(x => x.Index == i);
                if (urlOverlay != null && !string.IsNullOrEmpty(urlOverlay.OverlayName))
                    AddToList(ConvertToBase64UrlString(urlOverlay.OverlayName));
                else if (i != max)
                    AddEmpty();
            }

            if (IsEmpty()) return "";

            return "p.on=" + GetUrlValue();
        }

        private static string ConvertToBase64UrlString(string value)
        {
            return HttpUtility.UrlEncode(Convert.ToBase64String(Encoding.UTF8.GetBytes(value)));
        }
    }
}

File: sdk/js/PicarioXPO.RenderAPI.V2/XpoUrlGenerator/NameOverlayKey.ts
class NameOverlayKey extends OverlayKey {

    getValues(overlays: Array<XpoUrlOverlay>) {
        var max = UrlGeneratorModule.getMaxOverlayNumber(overlays);

        for (var i = 0; i <= max; i++) {
            var index = overlays.map(e => e.getIndex()).lastIndexOf(i);

            var urlOverlay = overlays[index];
            if (urlOverlay != null)
                this.addToList(this.convertToBase64UrlString(urlOverlay.getOverlayName()));
            else if (i != max)
                this.addEmpty();
        }

        if (this.isEmpty()) return "";

        return "p.on=" + this.getUrlValue();
    }

    convertToBase64UrlString(value: string) {
        var base64 = new Base64();

        return encodeURIComponent(base64.encode(value));
    }
}